src/xwrap.rs

```diff
@@ -2,340 +2,248 @@
+    pub fn root(&self) -> xproto::Window {
+        self.screen().root
+    }
 
-            let mut root = 0;
-            let mut parent = 0;
-            let mut children: *mut xlib::Window = ptr::null_mut();
-            let mut nchildren = 0;
-            xlib::XQueryTree(
-                self.handle,
-                window,
-                &mut root,
-                &mut parent,
-                &mut children,
-                &mut nchildren,
-            );
-            if !children.is_null() {
-                xlib::XFree(children as *mut raw::c_void);
-            }
+    pub fn get_window_geometry(&self, window: xproto::Window) -> Option<util::Rect> {
+        let geometry_cookie = self.conn.get_geometry(window).ok()?;
+        let tree_cookie = self.conn.query_tree(window).ok()?;
+        let geometry = geometry_cookie.reply().ok()?;
+        let tree = tree_cookie.reply().ok()?;
 
-            let mut x = attrs.x;
-            let mut y = attrs.y;
+        if tree.parent != 0 {
+            let cookie = self
+                .conn
+                .translate_coordinates(tree.parent, tree.root, geometry.x, geometry.y)
+                .ok()?;
+            let coords = cookie.reply().ok()?;
 
-            if parent != 0 {
-                let mut child = 0;
-                xlib::XTranslateCoordinates(
-                    self.handle,
-                    parent,
-                    root,
-                    attrs.x,
-                    attrs.y,
-                    &mut x,
-                    &mut y,
-                    &mut child,
-                );
-            }
+            Some(util::Rect {
+                x: coords.dst_x as i32,
+                y: coords.dst_y as i32,
+                w: geometry.width as i32,
+                h: geometry.height as i32,
+            })
+        } else {
+            Some(util::Rect {
+                x: geometry.x as i32,
+                y: geometry.y as i32,
+                w: geometry.width as i32,
+                h: geometry.height as i32,
+            })
+        }
+    }
 
-            util::Rect {
-                x,
-                y,
-                w: attrs.width,
-                h: attrs.height,
+    fn find_visual(&self, id: xproto::Visualid) -> Option<&xproto::Visualtype> {
+        for screen in &self.conn.setup().roots {
+            for depth in &screen.allowed_depths {
+                for visual in &depth.visuals {
+                    if visual.visual_id == id {
+                        return Some(visual);
+                    }
+                }
             }
         }
+        None
     }
 
-    pub fn get_image(&self, window: xlib::Window, rect: util::Rect) -> Option<Image> {
-        unsafe {
-            let all_planes = !0;
-            let image = xlib::XGetImage(
-                self.handle,
+    pub fn get_image(&self, window: xproto::Window, rect: util::Rect) -> Option<Image> {
+        let all_planes = !0;
```

Comment:
Why not a `const`?

Reply:
It used to be one, but this is really only relevant in this one place so I don't think it belongs in the global scope. The binding is only really there for code readability, passing `!0` directly to the function along with a comment would also work.

Reply:
My point was it could be
```suggestion
        const ALL_PLANES = !0;
```
even if it remains inline. But to be fair it's not a big deal.